Layer normalization fuse pass.

arogowie-intel opened the pull request:
### PR types
New features

### PR changes
 Others

### Describe
This PR introduces a fuse pass which converts the subgraph expressing layer normalization operation into paddle LayerNormalization op. The pattern is set as present in MLPerf BERT ONNX model converted to Paddle format.

The pattern from converted model looks like this:
```
/*
   *            X
   *           / \
   *          /   reduce_mean "u(x)"
   *          \   /
   *      elementwise_sub     "x - u(x)"
   *      /           \    2
   *      |            \  /
   *      |      elementwise_pow  "(x - u(x))^2"
   *      |             |
   *      |       reduce_mean     "sigma^2 = 1/C*Sum{(x - u(x))^2}"
   *      |             |     eps
   *      |             |     /
   *      |       elementwise_add "sigma^2 + epsilon"
   *      \             |
   *       \           sqrt       "sqrt(sigma^2 + epsilon)"
   *        \          /
   *         \        /
   *       elementwise_div        "lnorm = {x-u(x)}/{sqrt(sigma^2 + epsilon)}"
   *              |
   *       gamma  |
   *          \   |
   *       elementwise_mul        "scale: gamma(C) * lnorm"
   *              |
   *        beta  |
   *          \   |
   *       elementwise_add        "shift: gamma(C) * lnorm + beta(C)"
   */
   ```

arogowie-intel commented:
@wojtuss @jczaja @lidanqing-intel  Please take a look.

lidanqing-vv commented on python/paddle/fluid/tests/unittests/ir/inference/test_layer_norm_fuse_pass.py:
```diff
@@ -0,0 +1,63 @@
+from paddle.fluid.core import PassVersionChecker
+
+
+class LayerNormFusePassTest(InferencePassTest):
```
You could add a conv before, otherwise the model cannot be saved because no parameters need to be saved, so load_combine fail

```
class LayerNormFusePassTest(InferencePassTest):
    def setUp(self):
        with fluid.program_guard(self.main_program, self.startup_program):
            data = fluid.data(name="data", shape=[-1, 64, 12, 12], dtype="float32")
            conv_out = fluid.layers.conv2d(
                input=data,
                num_filters=6,
                filter_size=6,
                groups=1,
                padding=[1, 1, 1, 1],
                bias_attr=False,
                act=None)
            sqr_pow = fluid.layers.fill_constant(
                shape=[1], value=2, dtype="float32")
            eps = fluid.layers.fill_constant(
                shape=[1], value=1e-5, dtype="float32")
            gamma = fluid.data(name="gamma", shape=[9], dtype="float32")
            beta = fluid.data(name="beta", shape=[9], dtype="float32")

            x_mean_out = fluid.layers.mean(conv_out)
            x_sub_mean_out = fluid.layers.elementwise_sub(conv_out, x_mean_out)
            x_sub_mean_sqr_out = fluid.layers.elementwise_pow(x_sub_mean_out,
                                                              sqr_pow)
            std_dev_out = fluid.layers.mean(x_sub_mean_sqr_out)
            std_dev_eps_out = fluid.layers.elementwise_add(std_dev_out, eps)
            std_dev_eps_sqrt_out = fluid.layers.sqrt(std_dev_eps_out)
            division_out = fluid.layers.elementwise_div(x_sub_mean_out,
                                                        std_dev_eps_sqrt_out)
            scale_out = fluid.layers.elementwise_mul(division_out, gamma)
            shift_out = fluid.layers.elementwise_add(scale_out, beta)

        self.feeds = {
            "data": np.random.random((3, 64, 12, 12)).astype("float32"),
            "gamma": np.random.random((9)).astype("float32"),
            "beta": np.random.random((9)).astype("float32"),
        }
        self.fetch_list = [shift_out]

    def test_check_output(self):
        use_gpu = False
        self.check_output_with_option(use_gpu)
        self.assertTrue(PassVersionChecker.IsCompatible("layer_norm_fuse_pass"))
```

arogowie-intel replied:
The solution is to use [`create_parameter`](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/layers/tensor.py#L76) for gamma and beta.

lidanqing-vv commented on paddle/fluid/framework/ir/layer_norm_fuse_pass.cc:
```diff
@@ -0,0 +1,168 @@
+REGISTER_PASS_CAPABILITY(layer_norm_fuse_pass)
+    .AddCombination(
+        paddle::framework::compatible::OpVersionComparatorCombination()
+            .EQ("elementwise_add", 0)
```
.EQ("elementwise_add", 1)
            .EQ("elementwise_div", 1)
            .EQ("elementwise_mul", 1)
            .EQ("elementwise_pow", 1)
            .EQ("elementwise_sub", 1)
            .EQ("reduce_mean", 0)
            .EQ("sqrt", 0));


As we said, if one op has one checkpoint like following, its version increase 1. And why sqrt is here, sqrt is not an op I could not find this op
```
REGISTER_OP_VERSION(elementwise_pow)
    .AddCheckpoint(
        R"ROC(Register elementwise_pow for adding the attribute of Scale_y)ROC",
        paddle::framework::compatible::OpVersionDesc().NewAttr(
            "Scale_y",
```

arogowie-intel replied:
I've corrected the `EQ` condition to `GE` - now it works. Nor do I can find the kernel for this op :) However it is somehow automatically generated: https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/fluid/layers/ops.py#L40

lidanqing-vv replied:
If keep  .GE("elementwise_add", 0), for any update of elementwise_add, elementwise_sub op etc, layer_norm_fuse will not detect the changes. What if some attribute name changed etc, I am not sure, if you think no risks then it's fine

arogowie-intel replied:
@lidanqing-intel I've updated version conditions since I've added validating `reduce_mean` attributes. Please take a look.

lidanqing-intel reviewed (commented):
Add a op that has parameters to save before

lidanqing-intel commented on paddle/fluid/framework/ir/layer_norm_fuse_pass_tester.cc:
```diff
@@ -0,0 +1,293 @@
+                                      {"elementwise_div", 0},
+                                      {"elementwise_mul", 0},
+                                      {"layer_norm", 1}});
+}
```
Here also should be
```
  return test::AssertOpsCount(graph, {{"reduce_mean", 0},
                                      {"elementwise_sub", 1},
                                      {"elementwise_pow", 1},
                                      {"elementwise_add", 1},
                                      {"sqrt", 0},
                                      {"elementwise_div", 1},
                                      {"elementwise_mul", 1},
                                      {"layer_norm", 0}});
```
layer_norm version should be 0, because you did add checkpoint like this for it, so version 0.
```
REGISTER_OP_VERSION(elementwise_pow)
    .AddCheckpoint(
        R"ROC(Register elementwise_pow for adding the attribute of Scale_y)ROC",
        paddle::framework::compatible::OpVersionDesc().NewAttr(
```

jczaja commented on paddle/fluid/framework/ir/pass_test_util.cc:
```diff
@@ -168,6 +174,49 @@ bool RunPassAndAssert(Graph* graph, const std::string& pass_name,
 }
 
+template <typename T>
+void InitLoDTensorHolder(Scope* scope, const paddle::platform::Place& place,
```
Is this function used anywhere?

jczaja reviewed (dismissed):
LGTM

lidanqing-intel reviewed (approved):
LGTM

lidanqing-vv commented:
```
2021-01-29 21:55:16 0. You must have Dianhai approval for change 20+ files or add than 1000+ lines of content.
2021-01-29 21:55:16 1. You must have one RD (zhouwei25 (Recommend), luotao1) approval for tools/parallel_UT_rule.py changes, which manages the rule of running unittest with a same GPU. If the unittest failed due to Insufficient GPU memory or CUBLAS_STATUS_ALLOC_FAILED, you can remove it from tools/parallel_UT_rule.py.
2021-01-29 21:55:16 2. The error message you wrote in PADDLE_ENFORCE{_**} or PADDLE_THROW does not meet our error message writing specification. Possible errors include 1. the error message is empty / 2. the error message is too short / 3. the error type is not specified. Please read the specification [ https://github.com/PaddlePaddle/Paddle/wiki/Paddle-Error-Message-Writing-Specification ], then refine the error message. If it is a mismatch, please request chenwhql (Recommend), luotao1 or lanxianghit review and approve.
```

lidanqing-vv commented:
@arogowie-intel You can split some tests to next PR, and modify PADDLE_ENFORCE etc

arogowie-intel commented:
@luotao1 Could you start your review? 

* I've moved some of UTs to another PR (which I'll submit after this one is merged )
* The errors related to `PADDLE_ENFORCE` in PR-CI-APPROVAL are probably a mismatch. Maybe because they're fromatted into multiple lines?
```
2021-02-02 18:14:52 PADDLE_ENFORCE_NOT_NULL(graph, + platform::errors::InvalidArgument( + "The input graph of " + "LayerNormFusePass should not be nullptr.")); 
2021-02-02 18:14:52 PADDLE_ENFORCE_EQ(gamma_shape.size(), 1, + platform::errors::InvalidArgument( + "The LayerNorm gamma " + "(scale) tensor shape must be one-dimensional, " + "but is %s.", + gamma_shape.size())); 
2021-02-02 18:14:52 PADDLE_ENFORCE_EQ(beta_shape.size(), 1, + platform::errors::InvalidArgument( + "The LayerNorm beta " + "(shift) tensor shape must be one-dimensional, " + "but is %s.", + beta_shape.size())); 
2021-02-02 18:14:52 PADDLE_ENFORCE_EQ(beta_shape, gamma_shape, + platform::errors::InvalidArgument( + "The LayerNorm beta " + "and gamma tensors shapes' must be equal.")); 
2021-02-02 18:14:52 PADDLE_ENFORCE_EQ(gamma_shape.front(), x_last_dim, + platform::errors::InvalidArgument( + "The LayerNorm beta " + "and gamma tensors shapes' must be equal to the last " + "input's dimension size.")); 
```

luotao1 commented:
> Maybe because they're fromatted into multiple lines?

Got it.

```
2021-02-02 19:07:22 0. You must have one RD (zhouwei25 (Recommend), luotao1) approval for tools/parallel_UT_rule.py changes, which manages the rule of running unittest with a same GPU. If the unittest failed due to Insufficient GPU memory or CUBLAS_STATUS_ALLOC_FAILED, you can remove it from tools/parallel_UT_rule.py.
```
But why do you change `parallel_UT_rule.py`?

arogowie-intel commented:
> But why do you change `parallel_UT_rule.py`?

I've added there one UT, but got some problems with it then. So I removed it from this file. However my editor has removed trailing whitespaces in this file. I can revert this change.

wojtuss commented on paddle/fluid/framework/ir/layer_norm_fuse_pass.cc:
```diff
@@ -0,0 +1,231 @@
+    auto* eps_tensor = scope->FindVar(eps->Name())->GetMutable<LoDTensor>();
+
+    // ------------------ subgraph node's validation ---------------------------
+    PADDLE_ENFORCE_EQ(
```
It's a general comment: in fuse passes, if any condition precludes fusion, we try not to throw exceptions. Instead, we just skip fusing with appropriate message printed to the output. Please see if you can follow this approach in the fuse.

arogowie-intel replied:
Good point!

luotao1 reviewed (approved):
LGTM
